content/blog/prototyping/index.md

```diff
@@ -0,0 +1,767 @@
+This way, you can get *almost* the same experience as with a REPL in, say, Python or Ruby.
+
+The setup is simple:
+
+```sh
+# Install bacon
+cargo install --locked bacon
+
+# Run bacon in your project directory
+bacon
+```
+
+And just like that, you can get some pretty compilation output alongside your code editor.
+
+![bacon](bacon.png)
+
+Oh, and in case you were wondering, `cargo-watch` was another popular tool for
+this purpose, but it's since been [deprecated](https://github.com/watchexec/cargo-watch#maintenance).
+
+### `cargo-script` is awesome
+
+Did you know that cargo can also run scripts?
+
+For example, put this into a file called `script.rs`:
+
+```rust
+#!/usr/bin/env cargo +nightly -Zscript
+
+fn main() {
+    println!("Hello prototyping world");
+}
+```
+
+Now you can make the file executable with `chmod +x script.rs` and run it with `./script.rs` which it will compile and execute your code!
+This allows you to quickly test out ideas without having to create a new project.
+There is support for dependencies as well.
+
+At the moment, `cargo-script` is a nightly feature, but it will be released soon on stable Rust.
+You can read more about it in the [RFC](https://rust-lang.github.io/rfcs/3424-cargo-script.html).
+
+### Don't Worry About Performance
+
+You have to try really really hard to write slow code in Rust.
+Use that to your advantage: during the prototype phase, try to keep the code as simple as possible.
+
+I gave a talk titled ["The Four Horsemen of Bad Rust Code"](https://github.com/corrode/four-horsemen-talk) where I
+argue that premature optimization is one of the biggest sins in Rust. 
+
+Especially experienced developers coming from C or C++ are tempted to optimize too early.
+
+Rust makes code perform well by default - you get memory safety at virtually zero runtime cost. When developers try to optimize too early, they often run up against the borrow checker by using complex lifetime annotations and intricate reference patterns in pursuit of better performance.
+This leads to harder-to-maintain code that may not actually run faster.
+
+Resist the urge to optimize too early!
+You will thank yourself later. [^1]
+
+[^1]: In the talk, I show an example where early over-optimization led to the wrong abstraction and made the code slower. The actual bottleneck was elsewhere and hard to uncover without profiling.
+
+### Use `println!` and `dbg!` for debugging 
+
+I find that printing values is pretty handy while prototyping.
+It's one less context switch to make compared to starting a debugger.
+
+Most people use `println!` for that, but [`dbg!`](https://doc.rust-lang.org/std/macro.dbg.html) has a few advantages:
+
+- It prints the file name and line number where the macro is called. This helps you quickly find the source of the output.
+- It outputs the expression as well as its value.
+- It's less syntax-heavy than `println!`; e.g. `dbg!(x)` vs. `println!("{x:?}")`.
```

Comment:
The equivalent to `dbg!` is actually `println!("{x:#?}")` (which pretty-prints the debug output). Might be a bit too pedantic though. Mentioning it here would likely require at least a footnote.